<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title> 17 Unsupervised Learning | Exam PA Study Guide, Spring 2021</title>
  <meta name="description" content=" 17 Unsupervised Learning | Exam PA Study Guide, Spring 2021" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content=" 17 Unsupervised Learning | Exam PA Study Guide, Spring 2021" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content=" 17 Unsupervised Learning | Exam PA Study Guide, Spring 2021" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/artificial_actuary_logo_favicon.png" type="image/x-icon" />
<link rel="prev" href="tree-based-models.html"/>
<link rel="next" href="writing-and-communication.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Exam PA Study Manual</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#faq-frequently-asked-questions"><i class="fa fa-check"></i><b>0.1</b> FAQ: Frequently Asked Questions</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="the-exam.html"><a href="the-exam.html"><i class="fa fa-check"></i><b>1</b> The exam</a></li>
<li class="chapter" data-level="2" data-path="prometric-demo.html"><a href="prometric-demo.html"><i class="fa fa-check"></i><b>2</b> Prometric Demo</a></li>
<li class="chapter" data-level="3" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>3</b> Introduction</a></li>
<li class="chapter" data-level="4" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>4</b> Getting started</a>
<ul>
<li class="chapter" data-level="4.1" data-path="getting-started.html"><a href="getting-started.html#installing-r"><i class="fa fa-check"></i><b>4.1</b> Installing R</a></li>
<li class="chapter" data-level="4.2" data-path="getting-started.html"><a href="getting-started.html#installing-rstudio"><i class="fa fa-check"></i><b>4.2</b> Installing RStudio</a></li>
<li class="chapter" data-level="4.3" data-path="getting-started.html"><a href="getting-started.html#download-the-data"><i class="fa fa-check"></i><b>4.3</b> Download the data</a></li>
<li class="chapter" data-level="4.4" data-path="getting-started.html"><a href="getting-started.html#download-islr"><i class="fa fa-check"></i><b>4.4</b> Download ISLR</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="how-much-r-do-i-need-to-know-to-pass.html"><a href="how-much-r-do-i-need-to-know-to-pass.html"><i class="fa fa-check"></i><b>5</b> How much R do I need to know to pass?</a>
<ul>
<li class="chapter" data-level="5.1" data-path="how-much-r-do-i-need-to-know-to-pass.html"><a href="how-much-r-do-i-need-to-know-to-pass.html#how-to-use-the-pa-r-cheat-sheets"><i class="fa fa-check"></i><b>5.1</b> How to use the PA R cheat sheets?</a></li>
<li class="chapter" data-level="5.2" data-path="how-much-r-do-i-need-to-know-to-pass.html"><a href="how-much-r-do-i-need-to-know-to-pass.html#example-soa-pa-61620-task-8"><i class="fa fa-check"></i><b>5.2</b> Example: SOA PA 6/16/20, Task 8</a></li>
<li class="chapter" data-level="5.3" data-path="how-much-r-do-i-need-to-know-to-pass.html"><a href="how-much-r-do-i-need-to-know-to-pass.html#example-2---data-exploration"><i class="fa fa-check"></i><b>5.3</b> Example 2 - Data exploration</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="r-programming.html"><a href="r-programming.html"><i class="fa fa-check"></i><b>6</b> R programming</a>
<ul>
<li class="chapter" data-level="6.1" data-path="r-programming.html"><a href="r-programming.html#notebook-chunks"><i class="fa fa-check"></i><b>6.1</b> Notebook chunks</a></li>
<li class="chapter" data-level="6.2" data-path="r-programming.html"><a href="r-programming.html#basic-operations"><i class="fa fa-check"></i><b>6.2</b> Basic operations</a></li>
<li class="chapter" data-level="6.3" data-path="r-programming.html"><a href="r-programming.html#lists"><i class="fa fa-check"></i><b>6.3</b> Lists</a></li>
<li class="chapter" data-level="6.4" data-path="r-programming.html"><a href="r-programming.html#functions"><i class="fa fa-check"></i><b>6.4</b> Functions</a></li>
<li class="chapter" data-level="6.5" data-path="r-programming.html"><a href="r-programming.html#data-frames"><i class="fa fa-check"></i><b>6.5</b> Data frames</a></li>
<li class="chapter" data-level="6.6" data-path="r-programming.html"><a href="r-programming.html#pipes"><i class="fa fa-check"></i><b>6.6</b> Pipes</a></li>
<li class="chapter" data-level="6.7" data-path="r-programming.html"><a href="r-programming.html#the-code-of-soa-does-not-use-pipes-or-dplyr-so-can-i-skip-learning-this"><i class="fa fa-check"></i><b>6.7</b> The code of SOA does not use pipes or dplyr, so can I skip learning this?</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-exploration.html"><a href="data-exploration.html"><i class="fa fa-check"></i><b>7</b> Data exploration</a>
<ul>
<li class="chapter" data-level="7.1" data-path="data-exploration.html"><a href="data-exploration.html#how-to-make-graphs-in-r"><i class="fa fa-check"></i><b>7.1</b> How to make graphs in R?</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="data-exploration.html"><a href="data-exploration.html#add-a-plot"><i class="fa fa-check"></i><b>7.1.1</b> Add a plot</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="data-exploration.html"><a href="data-exploration.html#the-different-graph-types"><i class="fa fa-check"></i><b>7.2</b> The different graph types</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="data-exploration.html"><a href="data-exploration.html#histogram"><i class="fa fa-check"></i><b>7.2.1</b> Histogram</a></li>
<li class="chapter" data-level="7.2.2" data-path="data-exploration.html"><a href="data-exploration.html#box-plot"><i class="fa fa-check"></i><b>7.2.2</b> Box plot</a></li>
<li class="chapter" data-level="7.2.3" data-path="data-exploration.html"><a href="data-exploration.html#scatterplot"><i class="fa fa-check"></i><b>7.2.3</b> Scatterplot</a></li>
<li class="chapter" data-level="7.2.4" data-path="data-exploration.html"><a href="data-exploration.html#bar-charts"><i class="fa fa-check"></i><b>7.2.4</b> Bar charts</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="data-exploration.html"><a href="data-exploration.html#how-to-save-time-with-dplyr"><i class="fa fa-check"></i><b>7.3</b> How to save time with dplyr?</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="data-exploration.html"><a href="data-exploration.html#data-manipulation-chaining"><i class="fa fa-check"></i><b>7.3.1</b> Data manipulation chaining</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="data-exploration.html"><a href="data-exploration.html#how-to-explore-the-data"><i class="fa fa-check"></i><b>7.4</b> How to explore the data?</a></li>
<li class="chapter" data-level="7.5" data-path="data-exploration.html"><a href="data-exploration.html#how-to-transform-the-data"><i class="fa fa-check"></i><b>7.5</b> How to transform the data?</a></li>
<li class="chapter" data-level="7.6" data-path="data-exploration.html"><a href="data-exploration.html#missing-values"><i class="fa fa-check"></i><b>7.6</b> Missing values</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="data-exploration.html"><a href="data-exploration.html#types-of-missing-values"><i class="fa fa-check"></i><b>7.6.1</b> Types of Missing Values</a></li>
<li class="chapter" data-level="7.6.2" data-path="data-exploration.html"><a href="data-exploration.html#missing-value-resolutions"><i class="fa fa-check"></i><b>7.6.2</b> Missing Value Resolutions:</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-exploration.html"><a href="data-exploration.html#example-soa-pa-121219-task-1"><i class="fa fa-check"></i><b>7.7</b> Example: SOA PA 12/12/19, Task 1</a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="data-exploration.html"><a href="data-exploration.html#garbage-in-garbage-out"><i class="fa fa-check"></i><b>7.7.1</b> Garbage in; garbage out üóë</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-exploration.html"><a href="data-exploration.html#be-a-detective"><i class="fa fa-check"></i><b>7.7.2</b> Be a detective üîé</a></li>
<li class="chapter" data-level="7.7.3" data-path="data-exploration.html"><a href="data-exploration.html#a-picture-is-worth-a-thousand-words"><i class="fa fa-check"></i><b>7.7.3</b> A picture is worth a thousand words üì∑</a></li>
<li class="chapter" data-level="7.7.4" data-path="data-exploration.html"><a href="data-exploration.html#factor-or-numeric"><i class="fa fa-check"></i><b>7.7.4</b> Factor or numeric ‚ùì</a></li>
<li class="chapter" data-level="7.7.5" data-path="data-exploration.html"><a href="data-exploration.html#of-statistics-are-false"><i class="fa fa-check"></i><b>7.7.5</b> 73.6% of statistics are false üò®</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="data-exploration.html"><a href="data-exploration.html#exercises"><i class="fa fa-check"></i><b>7.8</b> Exercises</a>
<ul>
<li class="chapter" data-level="7.8.1" data-path="data-exploration.html"><a href="data-exploration.html#data-exploration-practice"><i class="fa fa-check"></i><b>7.8.1</b> Data Exploration Practice</a></li>
<li class="chapter" data-level="7.8.2" data-path="data-exploration.html"><a href="data-exploration.html#dplyr-practice"><i class="fa fa-check"></i><b>7.8.2</b> Dplyr Practice</a></li>
</ul></li>
<li class="chapter" data-level="7.9" data-path="data-exploration.html"><a href="data-exploration.html#answers-to-exercises"><i class="fa fa-check"></i><b>7.9</b> Answers to exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html"><i class="fa fa-check"></i><b>8</b> Introduction to modeling</a>
<ul>
<li class="chapter" data-level="8.1" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#modeling-vocabulary"><i class="fa fa-check"></i><b>8.1</b> Modeling vocabulary</a></li>
<li class="chapter" data-level="8.2" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#modeling-notation"><i class="fa fa-check"></i><b>8.2</b> Modeling notation</a></li>
<li class="chapter" data-level="8.3" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#ordinary-least-squares-ols"><i class="fa fa-check"></i><b>8.3</b> Ordinary Least Squares (OLS)</a></li>
<li class="chapter" data-level="8.4" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#r2-statistic"><i class="fa fa-check"></i><b>8.4</b> R^2 Statistic</a></li>
<li class="chapter" data-level="8.5" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#correlation"><i class="fa fa-check"></i><b>8.5</b> Correlation</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#pearsons-correlation"><i class="fa fa-check"></i><b>8.5.1</b> Pearson‚Äôs correlation</a></li>
<li class="chapter" data-level="8.5.2" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#spearman-rank-correlation"><i class="fa fa-check"></i><b>8.5.2</b> Spearman (rank) correlation</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#regression-vs.-classification"><i class="fa fa-check"></i><b>8.6</b> Regression vs.¬†classification</a></li>
<li class="chapter" data-level="8.7" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#regression-metrics"><i class="fa fa-check"></i><b>8.7</b> Regression metrics</a>
<ul>
<li class="chapter" data-level="8.7.1" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#example-soa-pa-61820-task-4"><i class="fa fa-check"></i><b>8.7.1</b> Example: SOA PA 6/18/20, Task 4</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="introduction-to-modeling.html"><a href="introduction-to-modeling.html#example-health-costs"><i class="fa fa-check"></i><b>8.8</b> Example: Health Costs</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="generalized-linear-models-glms.html"><a href="generalized-linear-models-glms.html"><i class="fa fa-check"></i><b>9</b> Generalized linear Models (GLMs)</a>
<ul>
<li class="chapter" data-level="9.0.1" data-path="generalized-linear-models-glms.html"><a href="generalized-linear-models-glms.html#assumptions-of-ols"><i class="fa fa-check"></i><b>9.0.1</b> Assumptions of OLS</a></li>
<li class="chapter" data-level="9.0.2" data-path="generalized-linear-models-glms.html"><a href="generalized-linear-models-glms.html#assumptions-of-glms"><i class="fa fa-check"></i><b>9.0.2</b> Assumptions of GLMs</a></li>
<li class="chapter" data-level="9.1" data-path="generalized-linear-models-glms.html"><a href="generalized-linear-models-glms.html#advantages-and-disadvantages"><i class="fa fa-check"></i><b>9.1</b> Advantages and disadvantages</a></li>
<li class="chapter" data-level="9.2" data-path="generalized-linear-models-glms.html"><a href="generalized-linear-models-glms.html#glms-for-regression"><i class="fa fa-check"></i><b>9.2</b> GLMs for regression</a></li>
<li class="chapter" data-level="9.3" data-path="generalized-linear-models-glms.html"><a href="generalized-linear-models-glms.html#interpretation-of-coefficients"><i class="fa fa-check"></i><b>9.3</b> Interpretation of coefficients</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="generalized-linear-models-glms.html"><a href="generalized-linear-models-glms.html#identity-link"><i class="fa fa-check"></i><b>9.3.1</b> Identity link</a></li>
<li class="chapter" data-level="9.3.2" data-path="generalized-linear-models-glms.html"><a href="generalized-linear-models-glms.html#log-link"><i class="fa fa-check"></i><b>9.3.2</b> Log link</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="generalized-linear-models-glms.html"><a href="generalized-linear-models-glms.html#other-links"><i class="fa fa-check"></i><b>9.4</b> Other links</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="glms-for-classification.html"><a href="glms-for-classification.html"><i class="fa fa-check"></i><b>10</b> GLMs for classification</a>
<ul>
<li class="chapter" data-level="10.1" data-path="glms-for-classification.html"><a href="glms-for-classification.html#binary-target"><i class="fa fa-check"></i><b>10.1</b> Binary target</a></li>
<li class="chapter" data-level="10.2" data-path="glms-for-classification.html"><a href="glms-for-classification.html#count-target"><i class="fa fa-check"></i><b>10.2</b> Count target</a></li>
<li class="chapter" data-level="10.3" data-path="glms-for-classification.html"><a href="glms-for-classification.html#link-functions"><i class="fa fa-check"></i><b>10.3</b> Link functions</a></li>
<li class="chapter" data-level="10.4" data-path="glms-for-classification.html"><a href="glms-for-classification.html#interpretation-of-coefficients-1"><i class="fa fa-check"></i><b>10.4</b> Interpretation of coefficients</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="glms-for-classification.html"><a href="glms-for-classification.html#logit"><i class="fa fa-check"></i><b>10.4.1</b> Logit</a></li>
<li class="chapter" data-level="10.4.2" data-path="glms-for-classification.html"><a href="glms-for-classification.html#probit-cauchit-cloglog"><i class="fa fa-check"></i><b>10.4.2</b> Probit, Cauchit, Cloglog</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="glms-for-classification.html"><a href="glms-for-classification.html#demo-the-model-for-interpretation"><i class="fa fa-check"></i><b>10.5</b> Demo the model for interpretation</a></li>
<li class="chapter" data-level="10.6" data-path="glms-for-classification.html"><a href="glms-for-classification.html#example---auto-claims"><i class="fa fa-check"></i><b>10.6</b> Example - Auto Claims</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="classification-metrics.html"><a href="classification-metrics.html"><i class="fa fa-check"></i><b>11</b> Classification metrics</a>
<ul>
<li class="chapter" data-level="11.1" data-path="classification-metrics.html"><a href="classification-metrics.html#area-under-the-roc-curve-auc"><i class="fa fa-check"></i><b>11.1</b> Area Under the ROC Curve (AUC)</a></li>
<li class="chapter" data-level="11.2" data-path="classification-metrics.html"><a href="classification-metrics.html#example---auto-claims-1"><i class="fa fa-check"></i><b>11.2</b> Example - Auto Claims</a></li>
<li class="chapter" data-level="11.3" data-path="classification-metrics.html"><a href="classification-metrics.html#example-soa-hr-task-5"><i class="fa fa-check"></i><b>11.3</b> Example: SOA HR, Task 5</a></li>
<li class="chapter" data-level="11.4" data-path="classification-metrics.html"><a href="classification-metrics.html#example-soa-pa-121219-task-11"><i class="fa fa-check"></i><b>11.4</b> Example: SOA PA 12/12/19, Task 11</a></li>
<li class="chapter" data-level="11.5" data-path="classification-metrics.html"><a href="classification-metrics.html#additional-reading"><i class="fa fa-check"></i><b>11.5</b> Additional reading</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html"><i class="fa fa-check"></i><b>12</b> Additional GLM topics</a>
<ul>
<li class="chapter" data-level="12.1" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#residuals"><i class="fa fa-check"></i><b>12.1</b> Residuals</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#raw-residuals"><i class="fa fa-check"></i><b>12.1.1</b> Raw residuals</a></li>
<li class="chapter" data-level="12.1.2" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#deviance-residuals"><i class="fa fa-check"></i><b>12.1.2</b> Deviance residuals</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#example"><i class="fa fa-check"></i><b>12.2</b> Example</a></li>
<li class="chapter" data-level="12.3" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#log-transforms-of-predictors"><i class="fa fa-check"></i><b>12.3</b> Log transforms of predictors</a></li>
<li class="chapter" data-level="12.4" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#example-1"><i class="fa fa-check"></i><b>12.4</b> Example</a></li>
<li class="chapter" data-level="12.5" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#reference-levels"><i class="fa fa-check"></i><b>12.5</b> Reference levels</a></li>
<li class="chapter" data-level="12.6" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#interactions"><i class="fa fa-check"></i><b>12.6</b> Interactions</a></li>
<li class="chapter" data-level="12.7" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#offsets"><i class="fa fa-check"></i><b>12.7</b> Offsets</a></li>
<li class="chapter" data-level="12.8" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#tweedie-regression"><i class="fa fa-check"></i><b>12.8</b> Tweedie regression</a></li>
<li class="chapter" data-level="12.9" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#combinations-of-link-functions-and-target-distributions"><i class="fa fa-check"></i><b>12.9</b> Combinations of Link Functions and Target Distributions</a>
<ul>
<li class="chapter" data-level="12.9.1" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#gaussian-response-with-log-link"><i class="fa fa-check"></i><b>12.9.1</b> Gaussian Response with Log Link</a></li>
<li class="chapter" data-level="12.9.2" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#gaussian-response-with-inverse-link"><i class="fa fa-check"></i><b>12.9.2</b> Gaussian Response with Inverse Link</a></li>
<li class="chapter" data-level="12.9.3" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#gaussian-response-with-identity-link"><i class="fa fa-check"></i><b>12.9.3</b> Gaussian Response with Identity Link</a></li>
<li class="chapter" data-level="12.9.4" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#gaussian-response-with-log-link-and-negative-values"><i class="fa fa-check"></i><b>12.9.4</b> Gaussian Response with Log Link and Negative Values</a></li>
<li class="chapter" data-level="12.9.5" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#gamma-response-with-log-link"><i class="fa fa-check"></i><b>12.9.5</b> Gamma Response with Log Link</a></li>
<li class="chapter" data-level="12.9.6" data-path="additional-glm-topics.html"><a href="additional-glm-topics.html#gamma-with-inverse-link"><i class="fa fa-check"></i><b>12.9.6</b> Gamma with Inverse Link</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html"><i class="fa fa-check"></i><b>13</b> GLM variable selection</a>
<ul>
<li class="chapter" data-level="13.1" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html#stepwise-subset-selection"><i class="fa fa-check"></i><b>13.1</b> Stepwise subset selection</a></li>
<li class="chapter" data-level="13.2" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html#example-soa-pa-61219-task-6"><i class="fa fa-check"></i><b>13.2</b> Example: SOA PA 6/12/19, Task 6</a></li>
<li class="chapter" data-level="13.3" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html#penalized-linear-models"><i class="fa fa-check"></i><b>13.3</b> Penalized Linear Models</a></li>
<li class="chapter" data-level="13.4" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html#ridge-regression"><i class="fa fa-check"></i><b>13.4</b> Ridge Regression</a></li>
<li class="chapter" data-level="13.5" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html#lasso"><i class="fa fa-check"></i><b>13.5</b> Lasso</a></li>
<li class="chapter" data-level="13.6" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html#elastic-net"><i class="fa fa-check"></i><b>13.6</b> Elastic Net</a></li>
<li class="chapter" data-level="13.7" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html#advantages-and-disadvantages-1"><i class="fa fa-check"></i><b>13.7</b> Advantages and disadvantages</a></li>
<li class="chapter" data-level="13.8" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html#example-ridge-regression"><i class="fa fa-check"></i><b>13.8</b> Example: Ridge Regression</a></li>
<li class="chapter" data-level="13.9" data-path="glm-variable-selection.html"><a href="glm-variable-selection.html#example-the-lasso"><i class="fa fa-check"></i><b>13.9</b> Example: The Lasso</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="bias-variance-trade-off.html"><a href="bias-variance-trade-off.html"><i class="fa fa-check"></i><b>14</b> Bias-variance trade-off</a></li>
<li class="chapter" data-level="15" data-path="visualization.html"><a href="visualization.html"><i class="fa fa-check"></i><b>15</b> Visualization</a>
<ul>
<li class="chapter" data-level="15.1" data-path="visualization.html"><a href="visualization.html#learning-outcomes"><i class="fa fa-check"></i><b>15.1</b> Learning Outcomes</a></li>
<li class="chapter" data-level="15.2" data-path="visualization.html"><a href="visualization.html#how-to-make-graphs-in-r-1"><i class="fa fa-check"></i><b>15.2</b> How to make graphs in R</a></li>
<li class="chapter" data-level="15.3" data-path="visualization.html"><a href="visualization.html#add-a-plot-1"><i class="fa fa-check"></i><b>15.3</b> Add a plot</a></li>
<li class="chapter" data-level="15.4" data-path="visualization.html"><a href="visualization.html#data-manipulation-chaining-1"><i class="fa fa-check"></i><b>15.4</b> Data manipulation chaining</a></li>
<li class="chapter" data-level="15.5" data-path="visualization.html"><a href="visualization.html#the-different-graph-types-1"><i class="fa fa-check"></i><b>15.5</b> The different graph types</a>
<ul>
<li class="chapter" data-level="15.5.1" data-path="visualization.html"><a href="visualization.html#histogram-1"><i class="fa fa-check"></i><b>15.5.1</b> Histogram</a></li>
<li class="chapter" data-level="15.5.2" data-path="visualization.html"><a href="visualization.html#box-plot-1"><i class="fa fa-check"></i><b>15.5.2</b> Box plot</a></li>
<li class="chapter" data-level="15.5.3" data-path="visualization.html"><a href="visualization.html#scatterplot-1"><i class="fa fa-check"></i><b>15.5.3</b> Scatterplot</a></li>
<li class="chapter" data-level="15.5.4" data-path="visualization.html"><a href="visualization.html#boxplot-univariate"><i class="fa fa-check"></i><b>15.5.4</b> Boxplot (Univariate)</a></li>
<li class="chapter" data-level="15.5.5" data-path="visualization.html"><a href="visualization.html#barplot-bivariate"><i class="fa fa-check"></i><b>15.5.5</b> Barplot (Bivariate)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="tree-based-models.html"><a href="tree-based-models.html"><i class="fa fa-check"></i><b>16</b> Tree-based models</a>
<ul>
<li class="chapter" data-level="16.1" data-path="tree-based-models.html"><a href="tree-based-models.html#decision-trees"><i class="fa fa-check"></i><b>16.1</b> Decision Trees</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="tree-based-models.html"><a href="tree-based-models.html#example-soa-pa-6182020-task-6"><i class="fa fa-check"></i><b>16.1.1</b> Example: SOA PA 6/18/2020, Task 6</a></li>
<li class="chapter" data-level="16.1.2" data-path="tree-based-models.html"><a href="tree-based-models.html#advantages-and-disadvantages-2"><i class="fa fa-check"></i><b>16.1.2</b> Advantages and disadvantages</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="tree-based-models.html"><a href="tree-based-models.html#ensemble-learning"><i class="fa fa-check"></i><b>16.2</b> Ensemble learning</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="tree-based-models.html"><a href="tree-based-models.html#bagging"><i class="fa fa-check"></i><b>16.2.1</b> Bagging</a></li>
<li class="chapter" data-level="16.2.2" data-path="tree-based-models.html"><a href="tree-based-models.html#boosting"><i class="fa fa-check"></i><b>16.2.2</b> Boosting</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="tree-based-models.html"><a href="tree-based-models.html#random-forests"><i class="fa fa-check"></i><b>16.3</b> Random Forests</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="tree-based-models.html"><a href="tree-based-models.html#example-2"><i class="fa fa-check"></i><b>16.3.1</b> Example</a></li>
<li class="chapter" data-level="16.3.2" data-path="tree-based-models.html"><a href="tree-based-models.html#variable-importance"><i class="fa fa-check"></i><b>16.3.2</b> Variable Importance</a></li>
<li class="chapter" data-level="16.3.3" data-path="tree-based-models.html"><a href="tree-based-models.html#partial-dependence"><i class="fa fa-check"></i><b>16.3.3</b> Partial dependence</a></li>
<li class="chapter" data-level="16.3.4" data-path="tree-based-models.html"><a href="tree-based-models.html#advantages-and-disadvantages-3"><i class="fa fa-check"></i><b>16.3.4</b> Advantages and disadvantages</a></li>
</ul></li>
<li class="chapter" data-level="16.4" data-path="tree-based-models.html"><a href="tree-based-models.html#gradient-boosted-trees"><i class="fa fa-check"></i><b>16.4</b> Gradient Boosted Trees</a>
<ul>
<li class="chapter" data-level="16.4.1" data-path="tree-based-models.html"><a href="tree-based-models.html#gradient-boosting"><i class="fa fa-check"></i><b>16.4.1</b> Gradient Boosting</a></li>
<li class="chapter" data-level="16.4.2" data-path="tree-based-models.html"><a href="tree-based-models.html#notation"><i class="fa fa-check"></i><b>16.4.2</b> Notation</a></li>
<li class="chapter" data-level="16.4.3" data-path="tree-based-models.html"><a href="tree-based-models.html#parameters"><i class="fa fa-check"></i><b>16.4.3</b> Parameters</a></li>
<li class="chapter" data-level="16.4.4" data-path="tree-based-models.html"><a href="tree-based-models.html#example-3"><i class="fa fa-check"></i><b>16.4.4</b> Example</a></li>
<li class="chapter" data-level="16.4.5" data-path="tree-based-models.html"><a href="tree-based-models.html#advantages-and-disadvantages-4"><i class="fa fa-check"></i><b>16.4.5</b> Advantages and disadvantages</a></li>
</ul></li>
<li class="chapter" data-level="16.5" data-path="tree-based-models.html"><a href="tree-based-models.html#exercises-1"><i class="fa fa-check"></i><b>16.5</b> Exercises</a>
<ul>
<li class="chapter" data-level="16.5.1" data-path="tree-based-models.html"><a href="tree-based-models.html#rf-tuning-with-caret"><i class="fa fa-check"></i><b>16.5.1</b> 1. RF tuning with <code>caret</code></a></li>
<li class="chapter" data-level="16.5.2" data-path="tree-based-models.html"><a href="tree-based-models.html#tuning-a-gbm-with-caret"><i class="fa fa-check"></i><b>16.5.2</b> 2. Tuning a GBM with <code>caret</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html"><i class="fa fa-check"></i><b>17</b> Unsupervised Learning</a>
<ul>
<li class="chapter" data-level="17.1" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#types-of-learning"><i class="fa fa-check"></i><b>17.1</b> Types of Learning</a></li>
<li class="chapter" data-level="17.2" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#correlation-analysis"><i class="fa fa-check"></i><b>17.2</b> Correlation Analysis</a>
<ul>
<li class="chapter" data-level="17.2.1" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#correlation-does-not-equal-causation"><i class="fa fa-check"></i><b>17.2.1</b> Correlation does not equal causation</a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>17.3</b> Principal Component Analysis (PCA)</a>
<ul>
<li class="chapter" data-level="17.3.1" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#example-us-arrests"><i class="fa fa-check"></i><b>17.3.1</b> Example: US Arrests</a></li>
<li class="chapter" data-level="17.3.2" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#example-soa-pa-61219-task-3"><i class="fa fa-check"></i><b>17.3.2</b> Example: SOA PA 6/12/19, Task 3</a></li>
<li class="chapter" data-level="17.3.3" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#example-pca-on-cancel-cells"><i class="fa fa-check"></i><b>17.3.3</b> Example: PCA on Cancel Cells</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#clustering"><i class="fa fa-check"></i><b>17.4</b> Clustering</a></li>
<li class="chapter" data-level="17.5" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#k-means-clustering"><i class="fa fa-check"></i><b>17.5</b> K-Means Clustering</a></li>
<li class="chapter" data-level="17.6" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#hierarchical-clustering"><i class="fa fa-check"></i><b>17.6</b> Hierarchical Clustering</a>
<ul>
<li class="chapter" data-level="17.6.1" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#example-clustering-cancel-cells"><i class="fa fa-check"></i><b>17.6.1</b> Example: Clustering Cancel Cells</a></li>
<li class="chapter" data-level="17.6.2" data-path="unsupervised-learning.html"><a href="unsupervised-learning.html#references"><i class="fa fa-check"></i><b>17.6.2</b> References</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="18" data-path="writing-and-communication.html"><a href="writing-and-communication.html"><i class="fa fa-check"></i><b>18</b> Writing and Communication</a>
<ul>
<li class="chapter" data-level="18.1" data-path="writing-and-communication.html"><a href="writing-and-communication.html#spelling-and-grammar"><i class="fa fa-check"></i><b>18.1</b> Spelling and Grammar</a></li>
<li class="chapter" data-level="18.2" data-path="writing-and-communication.html"><a href="writing-and-communication.html#how-to-write-for-pa"><i class="fa fa-check"></i><b>18.2</b> How to Write for PA</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i><b>19</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Exam PA Study Guide, Spring 2021</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="unsupervised-learning" class="section level1" number="17">
<h1><span class="header-section-number"> 17</span> Unsupervised Learning</h1>
<p>This chapter covers the eighth learning objective along with one additional topic of correlation analysis.</p>
<p><a href="https://exampa.net/"><img src="images/lesson-unsupervised-learning.PNG" alt="Lecture" /></a></p>
<p>Already enrolled? Watch the full video: <a target="_parent" href="https://course.exampa.net/mod/page/view.php?id=147">Practice Exams + Lessons</a></p>
<p><img src="images/learning_obj8.png" width="400%" style="display: block; margin: auto;" /></p>
<p>As of the fall of 2020, there have been eight different PA exams. The table below shows the types of questions in each. Your exam in December will likely have a question on either k-means or hierarchical clustering because these topics have not frequently been tested. The Hospital Readmissions sample project did have one question related to k-means, but that was in 2018. Our ExamPA.net practice exams contain questions on both topics.</p>
<p><img src="images/prior_unsupervised_qs.png" width="400%" style="display: block; margin: auto;" /></p>
<div id="types-of-learning" class="section level2" number="17.1">
<h2><span class="header-section-number">17.1</span> Types of Learning</h2>
<p>You are in a classroom and your teacher is giving a lesson which has a correct answer, such as 1 + 2 = 2. This is called <em>supervised learning</em>. The teacher is the supervisor, who is responsible for minimizing the number of incorrect answers. When a question is answered correctly, there is a reward, and when a mistake is made, there is a penalty. In machine learning, we measure the performance using metrics such as RMSE, and we say that the model is a good ‚Äúfit‚Äù when this metric is low.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-243"></span>
<img src="images/classroom.png" alt="Supervised (Left), Unsupervised (Right)" width="49%" height="20%" /><img src="images/legos.png" alt="Supervised (Left), Unsupervised (Right)" width="49%" height="20%" />
<p class="caption">
Figure 17.1: Supervised (Left), Unsupervised (Right)
</p>
</div>
<p>However, if you are given a pile of <strong>Legos</strong> and told to build whatever you want, that would be <em>unsupervised learning</em>. There are no right or wrong answers and to goal is to explore patterns in the data. There are no performance metrics, and it is impossible to say whether a model is a good fit or not.</p>
<p>Up to this point, you have been working with supervised learning. This is the major focus of PA because predictive analytics often is used to predict a future outcome. We now move on to the unsupervised learning algorithms. Here are all of the learning algorithms on PA:</p>
<table>
<thead>
<tr class="header">
<th>Supervised</th>
<th>Unsupervised</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>GLM</td>
<td>Correlation analysis</td>
</tr>
<tr class="even">
<td>Lasso, Ridge, and Elastic Net</td>
<td>Principal component analysis (PCA)</td>
</tr>
<tr class="odd">
<td>Decision Tree</td>
<td>K-means clustering</td>
</tr>
<tr class="even">
<td>Bagged Tree</td>
<td>Hierarchical clustering</td>
</tr>
<tr class="odd">
<td>Boosted Tree</td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Semi-Supervised Learning</strong> is a mix of the two. One example of this is using PCA or Clustering to create features used in a supervised model.</p>
</div>
<div id="correlation-analysis" class="section level2" number="17.2">
<h2><span class="header-section-number">17.2</span> Correlation Analysis</h2>
<ul>
<li>Two variables are said to be positively correlated when increasing one tends to increase the other and negatively correlated when increasing one decreases the other</li>
<li>Correlation is unsupervised because it does not depend on the target variable</li>
<li>Correlation is only defined for numeric variables. It is possible to calculate the correlation between binary variables if coded as 0 or 1, but these questions have never appeared on PA.</li>
</ul>
<div id="correlation-does-not-equal-causation" class="section level3" number="17.2.1">
<h3><span class="header-section-number">17.2.1</span> Correlation does not equal causation</h3>
<p>This is a common saying of statisticians. Two things happening at the same time are not sufficient evidence to suggest that one causes the other. A <a href="https://tylervigen.com/old-version.html">spurious correlation</a> is when two unrelated variables have a positive or negative correlation by random chance.</p>
<p>Examples:</p>
<p><strong>Number of Drownings and Ice Cream Sales:</strong> Drownings rise when ice cream sales rise because the heat causes more people to go swimming and want ice cream.</p>
<p><strong>Executives who say ‚Äúplease‚Äù and ‚Äúthank you‚Äù more often enjoy better performance reviews:</strong> It might at first appear that this is due to brown-nosing or flattery, but a more likely explanation is that people who take the extra effort to be polite also take the extra effort to do their jobs well. People who have good performance may also be polite, but not all polite people are high performers.</p>
</div>
</div>
<div id="principal-component-analysis-pca" class="section level2" number="17.3">
<h2><span class="header-section-number">17.3</span> Principal Component Analysis (PCA)</h2>
<p>Often there are variables that contain redundant information. PCA is one method of simplifying them.</p>
<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/FgakZw6K1QQ" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
</iframe>
<p>StatQuest. ‚ÄúPrincipal Component Analysis (PCA), Step-by-Step‚Äù YouTube, Joshua Starmer, 2 Apr 2018, <a href="https://www.youtube.com/watch?v=FgakZw6K1QQ" class="uri">https://www.youtube.com/watch?v=FgakZw6K1QQ</a></p>
<p>PCA is a <em>dimensionality reduction</em> method which reduces the number of variables needed to retain most of the information in a matrix. If there are predictor variables <span class="math inline">\((x_1, x_2, x_3)\)</span>, then running PCA and choosing the first two Principal Components (PCs) would reduce the dimension from 3 to 2. You can imagine this as observations in three-dimensional space being projected down onto a plane. The coordinates of this plane are <span class="math inline">\((\text{PC}_1, \text{PC}_2)\)</span>.</p>
<p><img src="images/3d_dim_reduction.png" width="392" style="display: block; margin: auto;" /></p>
<p>Each PC is a linear combination of the original variables. For example, PC1 might be</p>
<p><span class="math display">\[PC_1 = 0.2X_1 + 0.3X_2 - 0.2X_3\]</span>
The weights here are also called <em>loadings</em> or <em>rotations</em>, and are (0.2, 0.3, -0.2) in this example.</p>
<table>
<thead>
<tr class="header">
<th>Readings</th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ISLR 10.2 Principal Component Analysis</td>
<td></td>
</tr>
<tr class="even">
<td>ISLR 10.3 Clustering Methods</td>
<td></td>
</tr>
</tbody>
</table>
<div id="example-us-arrests" class="section level3" number="17.3.1">
<h3><span class="header-section-number">17.3.1</span> Example: US Arrests</h3>
<p>In this example, we perform PCA on the <code>USArrests</code> data set, which is part of
the base <code>R</code> package. The rows of the data set contain the 50 states, in
alphabetical order:</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb410-1"><a href="unsupervised-learning.html#cb410-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb410-2"><a href="unsupervised-learning.html#cb410-2" aria-hidden="true" tabindex="-1"></a>states<span class="ot">=</span><span class="fu">row.names</span>(USArrests)</span>
<span id="cb410-3"><a href="unsupervised-learning.html#cb410-3" aria-hidden="true" tabindex="-1"></a>states</span></code></pre></div>
<pre><code>##  [1] &quot;Alabama&quot;        &quot;Alaska&quot;         &quot;Arizona&quot;        &quot;Arkansas&quot;       &quot;California&quot;    
##  [6] &quot;Colorado&quot;       &quot;Connecticut&quot;    &quot;Delaware&quot;       &quot;Florida&quot;        &quot;Georgia&quot;       
## [11] &quot;Hawaii&quot;         &quot;Idaho&quot;          &quot;Illinois&quot;       &quot;Indiana&quot;        &quot;Iowa&quot;          
## [16] &quot;Kansas&quot;         &quot;Kentucky&quot;       &quot;Louisiana&quot;      &quot;Maine&quot;          &quot;Maryland&quot;      
## [21] &quot;Massachusetts&quot;  &quot;Michigan&quot;       &quot;Minnesota&quot;      &quot;Mississippi&quot;    &quot;Missouri&quot;      
## [26] &quot;Montana&quot;        &quot;Nebraska&quot;       &quot;Nevada&quot;         &quot;New Hampshire&quot;  &quot;New Jersey&quot;    
## [31] &quot;New Mexico&quot;     &quot;New York&quot;       &quot;North Carolina&quot; &quot;North Dakota&quot;   &quot;Ohio&quot;          
## [36] &quot;Oklahoma&quot;       &quot;Oregon&quot;         &quot;Pennsylvania&quot;   &quot;Rhode Island&quot;   &quot;South Carolina&quot;
## [41] &quot;South Dakota&quot;   &quot;Tennessee&quot;      &quot;Texas&quot;          &quot;Utah&quot;           &quot;Vermont&quot;       
## [46] &quot;Virginia&quot;       &quot;Washington&quot;     &quot;West Virginia&quot;  &quot;Wisconsin&quot;      &quot;Wyoming&quot;</code></pre>
<p>The columns of the data set contain four variables relating to various crimes:</p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb412-1"><a href="unsupervised-learning.html#cb412-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(USArrests)</span></code></pre></div>
<pre><code>## Rows: 50
## Columns: 4
## $ Murder   &lt;dbl&gt; 13.2, 10.0, 8.1, 8.8, 9.0, 7.9, 3.3, 5.9, 15.4, 17.4, 5.3, 2.6, 10.4, 7.2‚Ä¶
## $ Assault  &lt;int&gt; 236, 263, 294, 190, 276, 204, 110, 238, 335, 211, 46, 120, 249, 113, 56, ‚Ä¶
## $ UrbanPop &lt;int&gt; 58, 48, 80, 50, 91, 78, 77, 72, 80, 60, 83, 54, 83, 65, 57, 66, 52, 66, 5‚Ä¶
## $ Rape     &lt;dbl&gt; 21.2, 44.5, 31.0, 19.5, 40.6, 38.7, 11.1, 15.8, 31.9, 25.8, 20.2, 14.2, 2‚Ä¶</code></pre>
<p>Let us start by taking a quick look at the column means of the data.</p>
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb414-1"><a href="unsupervised-learning.html#cb414-1" aria-hidden="true" tabindex="-1"></a>USArrests <span class="sc">%&gt;%</span> <span class="fu">summarise_all</span>(mean)</span></code></pre></div>
<pre><code>##   Murder Assault UrbanPop   Rape
## 1  7.788  170.76    65.54 21.232</code></pre>
<p>We see right away the the data have <strong>vastly</strong> different means. We can also examine the variances of the four variables.</p>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb416-1"><a href="unsupervised-learning.html#cb416-1" aria-hidden="true" tabindex="-1"></a>USArrests <span class="sc">%&gt;%</span> <span class="fu">summarise_all</span>(var)</span></code></pre></div>
<pre><code>##     Murder  Assault UrbanPop     Rape
## 1 18.97047 6945.166 209.5188 87.72916</code></pre>
<p>Not surprisingly, the variables also have vastly different variances: the <code>UrbanPop</code> variable measures the percentage of the population in each state living in an urban area, which is not comparable to the number of crimes committed in each state per 100,000 individuals. If we failed to scale the variables before performing PCA, then most of the principal components we observed would be driven by the <code>Assault</code> variable, since it has by far the largest mean and variance.</p>
<p>Thus, it is important to standardize the variables to have mean zero and standard deviation 1 before performing PCA. We will perform principal components analysis using the <code>prcomp()</code> function, which is one of several functions that perform PCA. By default, this centers the variables to have a mean zero. By using the option <code>scale=TRUE</code>, we scale the variables to have standard
deviation 1:</p>
<div class="sourceCode" id="cb418"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb418-1"><a href="unsupervised-learning.html#cb418-1" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">=</span> <span class="fu">prcomp</span>(USArrests, <span class="at">scale=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>The output from <code>prcomp()</code> contains a number of useful quantities:</p>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb419-1"><a href="unsupervised-learning.html#cb419-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pca)</span></code></pre></div>
<pre><code>## [1] &quot;sdev&quot;     &quot;rotation&quot; &quot;center&quot;   &quot;scale&quot;    &quot;x&quot;</code></pre>
<p>The <code>center</code> and <code>scale</code> components correspond to the means and standard deviations of the variables that were used for scaling prior to implementing PCA:</p>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb421-1"><a href="unsupervised-learning.html#cb421-1" aria-hidden="true" tabindex="-1"></a>pca<span class="sc">$</span>center</span></code></pre></div>
<pre><code>##   Murder  Assault UrbanPop     Rape 
##    7.788  170.760   65.540   21.232</code></pre>
<div class="sourceCode" id="cb423"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb423-1"><a href="unsupervised-learning.html#cb423-1" aria-hidden="true" tabindex="-1"></a>pca<span class="sc">$</span>scale</span></code></pre></div>
<pre><code>##    Murder   Assault  UrbanPop      Rape 
##  4.355510 83.337661 14.474763  9.366385</code></pre>
<p>The rotation matrix provides the principal component loadings; each column
of <code>pr.out\$rotation</code> contains the corresponding principal component
loading vector:</p>
<div class="sourceCode" id="cb425"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb425-1"><a href="unsupervised-learning.html#cb425-1" aria-hidden="true" tabindex="-1"></a>pca<span class="sc">$</span>rotation</span></code></pre></div>
<pre><code>##                 PC1        PC2        PC3         PC4
## Murder   -0.5358995  0.4181809 -0.3412327  0.64922780
## Assault  -0.5831836  0.1879856 -0.2681484 -0.74340748
## UrbanPop -0.2781909 -0.8728062 -0.3780158  0.13387773
## Rape     -0.5434321 -0.1673186  0.8177779  0.08902432</code></pre>
<p>We see that there are four distinct principal components. This is to be
expected because there are in general <code>min(n ‚àí 1, p)</code> informative principal
components in a data set with <span class="math inline">\(n\)</span> observations and <span class="math inline">\(p\)</span> variables.</p>
<p>Using the <code>prcomp()</code> function, we do not need to explicitly multiply the data by the principal component loading vectors in order to obtain the principal component score vectors. Rather the 50 √ó 4 matrix <span class="math inline">\(x\)</span> has as its columns the principal component score vectors. That is, the $k^{th<code>column is the $k^{th</code> principal component score vector. We‚Äôll take a look at the first few states:</p>
<div class="sourceCode" id="cb427"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb427-1"><a href="unsupervised-learning.html#cb427-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pca<span class="sc">$</span>x)</span></code></pre></div>
<pre><code>##                   PC1        PC2         PC3          PC4
## Alabama    -0.9756604  1.1220012 -0.43980366  0.154696581
## Alaska     -1.9305379  1.0624269  2.01950027 -0.434175454
## Arizona    -1.7454429 -0.7384595  0.05423025 -0.826264240
## Arkansas    0.1399989  1.1085423  0.11342217 -0.180973554
## California -2.4986128 -1.5274267  0.59254100 -0.338559240
## Colorado   -1.4993407 -0.9776297  1.08400162  0.001450164</code></pre>
<p>We can plot the first two principal components using the <code>biplot()</code> function:</p>
<div class="sourceCode" id="cb429"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb429-1"><a href="unsupervised-learning.html#cb429-1" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(pca, <span class="at">scale=</span><span class="dv">0</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-254-1.png" width="768" /></p>
<p>The <code>scale=0</code> argument to <code>biplot()</code> ensures that the arrows are scaled to represent the loadings; other values for <code>scale</code> give slightly different bi plots with different interpretations.</p>
<p>The <code>prcomp()</code> function also outputs the standard deviation of each principal component. We can access these standard deviations as follows:</p>
<div class="sourceCode" id="cb430"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb430-1"><a href="unsupervised-learning.html#cb430-1" aria-hidden="true" tabindex="-1"></a>pca<span class="sc">$</span>sdev</span></code></pre></div>
<pre><code>## [1] 1.5748783 0.9948694 0.5971291 0.4164494</code></pre>
<p>The variance explained by each principal component is obtained by squaring
these:</p>
<div class="sourceCode" id="cb432"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb432-1"><a href="unsupervised-learning.html#cb432-1" aria-hidden="true" tabindex="-1"></a>pca_var<span class="ot">=</span>pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb432-2"><a href="unsupervised-learning.html#cb432-2" aria-hidden="true" tabindex="-1"></a>pca_var</span></code></pre></div>
<pre><code>## [1] 2.4802416 0.9897652 0.3565632 0.1734301</code></pre>
<p>To compute the proportion of variance explained by each principal component, we simply divide the variance explained by each principal component by the total variance explained by all four principal components:</p>
<div class="sourceCode" id="cb434"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb434-1"><a href="unsupervised-learning.html#cb434-1" aria-hidden="true" tabindex="-1"></a>pve<span class="ot">=</span>pca_var<span class="sc">/</span><span class="fu">sum</span>(pca_var)</span>
<span id="cb434-2"><a href="unsupervised-learning.html#cb434-2" aria-hidden="true" tabindex="-1"></a>pve</span></code></pre></div>
<pre><code>## [1] 0.62006039 0.24744129 0.08914080 0.04335752</code></pre>
<p>We see that the first principal component explains 62.0% of the variance in the data, the next principal component explains 24.7% of the variance, and so forth. We can plot the PVE explained by each component as follows:</p>
<div class="sourceCode" id="cb436"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb436-1"><a href="unsupervised-learning.html#cb436-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pve, <span class="at">xlab=</span><span class="st">&quot;Principal Component&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Proportion of Variance Explained&quot;</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>),<span class="at">type=</span><span class="st">&#39;b&#39;</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-258-1.png" width="576" /></p>
<p>We can also use the function <code>cumsum()</code>, which computes the cumulative sum of the elements of a numeric vector, to plot the cumulative PVE:</p>
<div class="sourceCode" id="cb437"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb437-1"><a href="unsupervised-learning.html#cb437-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pve), <span class="at">xlab=</span><span class="st">&quot;Principal Component&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Cumulative Proportion of Variance Explained&quot;</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>),<span class="at">type=</span><span class="st">&#39;b&#39;</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-259-1.png" width="576" /></p>
<div class="sourceCode" id="cb438"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb438-1"><a href="unsupervised-learning.html#cb438-1" aria-hidden="true" tabindex="-1"></a>a<span class="ot">=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">8</span>,<span class="sc">-</span><span class="dv">3</span>)</span>
<span id="cb438-2"><a href="unsupervised-learning.html#cb438-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cumsum</span>(a)</span></code></pre></div>
<pre><code>## [1]  1  3 11  8</code></pre>
</div>
<div id="example-soa-pa-61219-task-3" class="section level3" number="17.3.2">
<h3><span class="header-section-number">17.3.2</span> Example: SOA PA 6/12/19, Task 3</h3>
<blockquote>
<ol start="3" style="list-style-type: decimal">
<li>(9 points) Use observations from principal components analysis (PCA) to generate a new feature</li>
</ol>
</blockquote>
<blockquote>
<p>Your assistant has provided code to run a PCA on three variables. Run the code on these three variables. Interpret the output, including the loadings on significant principal components. Generate one new feature based on your observations (which may also involve dropping some current variables). Your assistant has provided some notes on using PCA on factor variables in the .Rmd file.</p>
</blockquote>
<iframe src="https://player.vimeo.com/video/465063620" width="640" height="360" frameborder="0" allow="autoplay; fullscreen" allowfullscreen>
</iframe>
<p>Already enrolled? Watch the full video: <a target="_parent" href="https://course.exampa.net/mod/page/view.php?id=204">Practice Exams</a> | <a target="_parent" href="https://course.exampa.net/mod/page/view.php?id=155">Practice Exams + Lessons</a></p>
</div>
<div id="example-pca-on-cancel-cells" class="section level3" number="17.3.3">
<h3><span class="header-section-number">17.3.3</span> Example: PCA on Cancel Cells</h3>
<p>The data <code>NCI60</code> contains expression levels of 6,830 genes from 64 cancel cell lines. Cancer type is also recorded.</p>
<div class="sourceCode" id="cb440"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb440-1"><a href="unsupervised-learning.html#cb440-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ISLR)</span>
<span id="cb440-2"><a href="unsupervised-learning.html#cb440-2" aria-hidden="true" tabindex="-1"></a>nci_labs<span class="ot">=</span>NCI60<span class="sc">$</span>labs</span>
<span id="cb440-3"><a href="unsupervised-learning.html#cb440-3" aria-hidden="true" tabindex="-1"></a>nci_data<span class="ot">=</span>NCI60<span class="sc">$</span>data</span></code></pre></div>
<p>We first perform PCA on the data after scaling the variables (genes) to have a standard deviation one, although one could reasonably argue that it is better not to scale the genes:</p>
<div class="sourceCode" id="cb441"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb441-1"><a href="unsupervised-learning.html#cb441-1" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">=</span> <span class="fu">prcomp</span>(nci_data, <span class="at">scale=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>We now plot the first few principal component score vectors in order to visualize the data. The observations (cell lines) corresponding to a given cancer type will be plotted in the same color to see to what extent the observations within a cancer type are similar to each other. We first create a simple function that assigns a distinct color to each element of a numeric vector. The function will be used to assign a color to each of the 64 cell lines, based on the cancer type to which it corresponds. We will use the <code>rainbow()</code> function, which takes its argument a positive integer and returns a vector containing that number of distinct colors.</p>
<div class="sourceCode" id="cb442"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb442-1"><a href="unsupervised-learning.html#cb442-1" aria-hidden="true" tabindex="-1"></a>Cols<span class="ot">=</span><span class="cf">function</span>(vec){</span>
<span id="cb442-2"><a href="unsupervised-learning.html#cb442-2" aria-hidden="true" tabindex="-1"></a>    cols<span class="ot">=</span><span class="fu">rainbow</span>(<span class="fu">length</span>(<span class="fu">unique</span>(vec)))</span>
<span id="cb442-3"><a href="unsupervised-learning.html#cb442-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(cols[<span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(vec))])</span>
<span id="cb442-4"><a href="unsupervised-learning.html#cb442-4" aria-hidden="true" tabindex="-1"></a>  }</span></code></pre></div>
<p>We now can plot the principal component score vectors:</p>
<div class="sourceCode" id="cb443"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb443-1"><a href="unsupervised-learning.html#cb443-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb443-2"><a href="unsupervised-learning.html#cb443-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col=</span><span class="fu">Cols</span>(nci_labs), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">&quot;Z1&quot;</span>,<span class="at">ylab=</span><span class="st">&quot;Z2&quot;</span>)</span>
<span id="cb443-3"><a href="unsupervised-learning.html#cb443-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca<span class="sc">$</span>x[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)], <span class="at">col=</span><span class="fu">Cols</span>(nci_labs), <span class="at">pch=</span><span class="dv">19</span>,<span class="at">xlab=</span><span class="st">&quot;Z1&quot;</span>,<span class="at">ylab=</span><span class="st">&quot;Z3&quot;</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-263-1.png" width="576" /></p>
<p>On the whole, cell lines corresponding to a single cancer type tend to have similar values on the first few principal component score vectors.</p>
<p>This indicates that cell lines from the same cancer type tend to have pretty similar gene expression levels.</p>
<p>We can obtain a summary of the proportion of variance explained (PVE) of the first few principal components using the (PVE) of the first few principal components using the <code>summary()</code> method for a <code>prcomp</code> object:</p>
<div class="sourceCode" id="cb444"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb444-1"><a href="unsupervised-learning.html#cb444-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pca)</span></code></pre></div>
<pre><code>## Importance of components:
##                            PC1      PC2      PC3      PC4      PC5      PC6      PC7
## Standard deviation     27.8535 21.48136 19.82046 17.03256 15.97181 15.72108 14.47145
## Proportion of Variance  0.1136  0.06756  0.05752  0.04248  0.03735  0.03619  0.03066
## Cumulative Proportion   0.1136  0.18115  0.23867  0.28115  0.31850  0.35468  0.38534
##                             PC8      PC9     PC10     PC11     PC12     PC13     PC14
## Standard deviation     13.54427 13.14400 12.73860 12.68672 12.15769 11.83019 11.62554
## Proportion of Variance  0.02686  0.02529  0.02376  0.02357  0.02164  0.02049  0.01979
## Cumulative Proportion   0.41220  0.43750  0.46126  0.48482  0.50646  0.52695  0.54674
##                            PC15     PC16     PC17     PC18     PC19    PC20     PC21
## Standard deviation     11.43779 11.00051 10.65666 10.48880 10.43518 10.3219 10.14608
## Proportion of Variance  0.01915  0.01772  0.01663  0.01611  0.01594  0.0156  0.01507
## Cumulative Proportion   0.56590  0.58361  0.60024  0.61635  0.63229  0.6479  0.66296
##                           PC22    PC23    PC24    PC25    PC26    PC27   PC28    PC29
## Standard deviation     10.0544 9.90265 9.64766 9.50764 9.33253 9.27320 9.0900 8.98117
## Proportion of Variance  0.0148 0.01436 0.01363 0.01324 0.01275 0.01259 0.0121 0.01181
## Cumulative Proportion   0.6778 0.69212 0.70575 0.71899 0.73174 0.74433 0.7564 0.76824
##                           PC30    PC31    PC32    PC33    PC34    PC35    PC36    PC37
## Standard deviation     8.75003 8.59962 8.44738 8.37305 8.21579 8.15731 7.97465 7.90446
## Proportion of Variance 0.01121 0.01083 0.01045 0.01026 0.00988 0.00974 0.00931 0.00915
## Cumulative Proportion  0.77945 0.79027 0.80072 0.81099 0.82087 0.83061 0.83992 0.84907
##                           PC38    PC39    PC40    PC41   PC42    PC43   PC44    PC45   PC46
## Standard deviation     7.82127 7.72156 7.58603 7.45619 7.3444 7.10449 7.0131 6.95839 6.8663
## Proportion of Variance 0.00896 0.00873 0.00843 0.00814 0.0079 0.00739 0.0072 0.00709 0.0069
## Cumulative Proportion  0.85803 0.86676 0.87518 0.88332 0.8912 0.89861 0.9058 0.91290 0.9198
##                           PC47    PC48    PC49    PC50    PC51    PC52    PC53    PC54
## Standard deviation     6.80744 6.64763 6.61607 6.40793 6.21984 6.20326 6.06706 5.91805
## Proportion of Variance 0.00678 0.00647 0.00641 0.00601 0.00566 0.00563 0.00539 0.00513
## Cumulative Proportion  0.92659 0.93306 0.93947 0.94548 0.95114 0.95678 0.96216 0.96729
##                           PC55    PC56    PC57   PC58    PC59    PC60    PC61    PC62
## Standard deviation     5.91233 5.73539 5.47261 5.2921 5.02117 4.68398 4.17567 4.08212
## Proportion of Variance 0.00512 0.00482 0.00438 0.0041 0.00369 0.00321 0.00255 0.00244
## Cumulative Proportion  0.97241 0.97723 0.98161 0.9857 0.98940 0.99262 0.99517 0.99761
##                           PC63      PC64
## Standard deviation     4.04124 1.237e-14
## Proportion of Variance 0.00239 0.000e+00
## Cumulative Proportion  1.00000 1.000e+00</code></pre>
<p>Using the <code>plot()</code> function, we can also plot the variance explained by the first few principal components:</p>
<div class="sourceCode" id="cb446"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb446-1"><a href="unsupervised-learning.html#cb446-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-265-1.png" width="576" /></p>
<p>Note that the height of each bar in the bar plot is given by squaring the corresponding element of <code>pr.out\$sdev</code>. However, it is generally more informative to plot the PVE of each principal component (i.e.¬†a <strong>scree plot</strong>) and the cumulative PVE of each principal component. This can be done with just a little tweaking:</p>
<div class="sourceCode" id="cb447"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb447-1"><a href="unsupervised-learning.html#cb447-1" aria-hidden="true" tabindex="-1"></a>pve <span class="ot">=</span> <span class="dv">100</span><span class="sc">*</span>pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span><span class="fu">sum</span>(pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb447-2"><a href="unsupervised-learning.html#cb447-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb447-3"><a href="unsupervised-learning.html#cb447-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pve,  <span class="at">type=</span><span class="st">&quot;o&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;PVE&quot;</span>, <span class="at">xlab=</span><span class="st">&quot;Principal Component&quot;</span>, <span class="at">col=</span><span class="st">&quot;blue&quot;</span>)</span>
<span id="cb447-4"><a href="unsupervised-learning.html#cb447-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">cumsum</span>(pve), <span class="at">type=</span><span class="st">&quot;o&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Cumulative PVE&quot;</span>, <span class="at">xlab=</span><span class="st">&quot;Principal Component&quot;</span>, <span class="at">col=</span><span class="st">&quot;brown3&quot;</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-266-1.png" width="576" /></p>
<p>Together, we see that the first seven principal components explain around 40% of the variance in the data. This is not a huge amount of the variance. However, looking at the scree plot, we see that while each of the first seven principal components explains a substantial amount of variance, there is a marked decrease in the variance explained by other principal components. That is, there is an <strong>elbow</strong> in the plot after approximately the seventh principal component. This suggests that there may be little benefit to examining more than seven or so principal components (phew! even examining seven principal components may be difficult).</p>
</div>
</div>
<div id="clustering" class="section level2" number="17.4">
<h2><span class="header-section-number">17.4</span> Clustering</h2>
<p>Imagine that you are a large retailer interested in understanding the customer base. There may be several ‚Äútypes‚Äù of customers, such as those shopping for a business with corporate accounts, those shopping for leisure, or debt-strapped grad students. Each of these customers would exhibit different behavior and should be treated differently statistically. However, how can the ‚Äútype‚Äù of customer be defined? Especially for large customer data sets in the millions, one can imagine how this problem can be challenging.</p>
<p>Clustering algorithms look for groups of observations that are similar to one another. Because there is no target variable, measuring the quality of the ‚Äúfit‚Äù is much more complicated. There are many clustering algorithms, but this exam only focuses on the two that are most common.</p>
</div>
<div id="k-means-clustering" class="section level2" number="17.5">
<h2><span class="header-section-number">17.5</span> K-Means Clustering</h2>
<p>Kmeans takes continuous data and assigns observations into k clusters or groups. In the two-dimensional example, this is the same as drawing lines around points.</p>
<iframe width="560" height="315" src="https://www.youtube.com/embed/4b5d3muPQmA" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
</iframe>
<p>StatQuest. ‚Äú K-means clustering‚Äù YouTube, Joshua Starmer, 2 Apr 2018, <a href="https://www.youtube.com/watch?v=4b5d3muPQmA&amp;t=268s" class="uri">https://www.youtube.com/watch?v=4b5d3muPQmA&amp;t=268s</a></p>
<p>Kmeans consists of the following steps:</p>
<p><strong>a)</strong> Start with two variables (<span class="math inline">\(X_1\)</span> on the X-axis, and <span class="math inline">\(X_2\)</span> on the Y-axis.)
<strong>b)</strong> Randomly assign cluster centers.
<strong>c)</strong> Put each observation into the cluster that is closest.
<strong>d) - f)</strong> Move the cluster center to the mean of the observations assigned to it and continue until the centers stop moving.
<strong>g)</strong> Repeated steps a) - f) a given number of times (controlled by <code>n.starts</code>). This reduces the uncertainty from choosing the initial centers randomly.</p>
<p><img src="images/kmeans.png" width="1000%" /></p>
<div id="r-example" class="section level4" number="17.5.0.1">
<h4><span class="header-section-number">17.5.0.1</span> R Example</h4>
<p>In <code>R</code>, the function <code>kmeans()</code> performs K-means clustering in R. We begin with a simple simulated example in which there truly are two clusters in the data: the first 25 observations have a mean shift relative to the next 25 observations.</p>
<div class="sourceCode" id="cb448"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb448-1"><a href="unsupervised-learning.html#cb448-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb448-2"><a href="unsupervised-learning.html#cb448-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(<span class="dv">50</span><span class="sc">*</span><span class="dv">2</span>), <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb448-3"><a href="unsupervised-learning.html#cb448-3" aria-hidden="true" tabindex="-1"></a>x[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>,<span class="dv">1</span>] <span class="ot">=</span> x[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>,<span class="dv">1</span>]<span class="sc">+</span><span class="dv">3</span></span>
<span id="cb448-4"><a href="unsupervised-learning.html#cb448-4" aria-hidden="true" tabindex="-1"></a>x[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>,<span class="dv">2</span>] <span class="ot">=</span> x[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>,<span class="dv">2</span>]<span class="sc">-</span><span class="dv">4</span></span></code></pre></div>
<p>We now perform K-means clustering with <code>K  =  2</code>:</p>
<div class="sourceCode" id="cb449"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb449-1"><a href="unsupervised-learning.html#cb449-1" aria-hidden="true" tabindex="-1"></a>km_out <span class="ot">=</span> <span class="fu">kmeans</span>(x,<span class="dv">2</span>,<span class="at">nstart =</span> <span class="dv">20</span>)</span></code></pre></div>
<p>The cluster assignments of the 50 observations are contained in
<code>km_out$cluster</code>:</p>
<div class="sourceCode" id="cb450"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb450-1"><a href="unsupervised-learning.html#cb450-1" aria-hidden="true" tabindex="-1"></a>km_out<span class="sc">$</span>cluster</span></code></pre></div>
<pre><code>##  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
## [45] 2 2 2 2 2 2</code></pre>
<p>The K-means clustering perfectly separated the observations into two clusters even though we did not supply any group information to <code>kmeans()</code>. We can plot the data, with each observation colored according to its cluster assignment:</p>
<div class="sourceCode" id="cb452"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb452-1"><a href="unsupervised-learning.html#cb452-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, <span class="at">col =</span> (km_out<span class="sc">$</span>cluster<span class="sc">+</span><span class="dv">1</span>), <span class="at">main =</span> <span class="st">&quot;K-Means Clustering Results with K = 2&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="at">pch =</span> <span class="dv">20</span>, <span class="at">cex =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-271-1.png" width="576" /></p>
<p>Here the observations can be easily plotted because they are two-dimensional. If there were more than two variables, we could instead perform PCA and plot the first two principal components score vectors.
In this example, we knew that there were two clusters because we generated the data. However, for real data, in general, we do not know the true number of clusters. We could instead have performed K-means clustering on this example with <code>K  =  3</code>. If we do this, K-means clustering will split up the two ‚Äúreal‚Äù clusters since it has no information about them:</p>
<div class="sourceCode" id="cb453"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb453-1"><a href="unsupervised-learning.html#cb453-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">4</span>)</span>
<span id="cb453-2"><a href="unsupervised-learning.html#cb453-2" aria-hidden="true" tabindex="-1"></a>km_out <span class="ot">=</span> <span class="fu">kmeans</span>(x, <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb453-3"><a href="unsupervised-learning.html#cb453-3" aria-hidden="true" tabindex="-1"></a>km_out</span></code></pre></div>
<pre><code>## K-means clustering with 3 clusters of sizes 17, 23, 10
## 
## Cluster means:
##         [,1]        [,2]
## 1  3.7789567 -4.56200798
## 2 -0.3820397 -0.08740753
## 3  2.3001545 -2.69622023
## 
## Clustering vector:
##  [1] 1 3 1 3 1 1 1 3 1 3 1 3 1 3 1 3 1 1 1 1 1 3 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 3
## [45] 2 3 2 2 2 2
## 
## Within cluster sum of squares by cluster:
## [1] 25.74089 52.67700 19.56137
##  (between_SS / total_SS =  79.3 %)
## 
## Available components:
## 
## [1] &quot;cluster&quot;      &quot;centers&quot;      &quot;totss&quot;        &quot;withinss&quot;     &quot;tot.withinss&quot;
## [6] &quot;betweenss&quot;    &quot;size&quot;         &quot;iter&quot;         &quot;ifault&quot;</code></pre>
<div class="sourceCode" id="cb455"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb455-1"><a href="unsupervised-learning.html#cb455-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, <span class="at">col =</span> (km_out<span class="sc">$</span>cluster<span class="sc">+</span><span class="dv">1</span>), <span class="at">main =</span> <span class="st">&quot;K-Means Clustering Results with K = 3&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="at">pch =</span> <span class="dv">20</span>, <span class="at">cex =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-272-1.png" width="576" /></p>
<p>To run the <code>kmeans()</code> function in R with multiple initial cluster assignments, we use the nstart argument. If a value of <code>nstart</code> greater than one is used, then K-means clustering will be performed using multiple random assignments, and the <code>kmeans()</code> function will report only the best results. Here we compare using <code>nstart = 1</code>:</p>
<div class="sourceCode" id="cb456"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb456-1"><a href="unsupervised-learning.html#cb456-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3</span>)</span>
<span id="cb456-2"><a href="unsupervised-learning.html#cb456-2" aria-hidden="true" tabindex="-1"></a>km_out <span class="ot">=</span> <span class="fu">kmeans</span>(x, <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">1</span>)</span>
<span id="cb456-3"><a href="unsupervised-learning.html#cb456-3" aria-hidden="true" tabindex="-1"></a>km_out<span class="sc">$</span>tot.withinss</span></code></pre></div>
<pre><code>## [1] 97.97927</code></pre>
<p>to <code>nstart = 20</code>:</p>
<div class="sourceCode" id="cb458"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb458-1"><a href="unsupervised-learning.html#cb458-1" aria-hidden="true" tabindex="-1"></a>km_out <span class="ot">=</span> <span class="fu">kmeans</span>(x,<span class="dv">3</span>,<span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb458-2"><a href="unsupervised-learning.html#cb458-2" aria-hidden="true" tabindex="-1"></a>km_out<span class="sc">$</span>tot.withinss</span></code></pre></div>
<pre><code>## [1] 97.97927</code></pre>
<p>Note that <code>km_out\$tot.withinss</code> is the total within-cluster sum of squares, which we seek to minimize by performing K-means clustering. The individual within-cluster sum-of-squares are contained in the vector <code>km_out\$withinss</code>.</p>
<p>It is generally recommended to always run K-means clustering with a large value of <code>nstart</code>, such as 20 or 50, to avoid getting stuck in an undesirable local optimum.</p>
<p>When performing K-means clustering and using multiple initial cluster assignments, it is also important to set a random seed using the
<code>set.seed()</code> function. This way, the initial cluster assignments can be replicated, and the K-means output will be fully reproducible.</p>
</div>
</div>
<div id="hierarchical-clustering" class="section level2" number="17.6">
<h2><span class="header-section-number">17.6</span> Hierarchical Clustering</h2>
<p>Kmeans required that we choose the number of clusters, k. Hierarchical clustering is an alternative that does not require that we choose only one value of k.</p>
<iframe width="560" height="315" src="https://www.youtube.com/embed/7xHsRkOdVwo" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
</iframe>
<p>StatQuest. ‚Äú Hierarchical Clustering‚Äù YouTube, Joshua Starmer, 20 Jun, 2017, <a href="https://www.youtube.com/watch?v=7xHsRkOdVwo&amp;t=137s" class="uri">https://www.youtube.com/watch?v=7xHsRkOdVwo&amp;t=137s</a></p>
<p>The most common type of hierarchical clustering uses a <em>bottom-up</em> approach. This starts with a single <strong>observation</strong> and then looks for others close and puts them into a cluster. Then it looks for other <strong>clusters</strong> that are similar and groups these together into a <strong>mega cluster</strong>. It continues to do this until all observations are in the same group.</p>
<p>This is analyzed with a graph called a dendrogram (dendro = tree, gram = graph). The height represents ‚Äúdistance,‚Äù or how similar the clusters are to one another. The clusters on the bottom, which are vertically close to one another, have similar data values; the clusters that are further apart vertically are less similar.</p>
<p>Choosing the value of the cutoff height changes the number of clusters that result.</p>
<p><img src="images/HClustering.png" width="1000%" style="display: block; margin: auto;" /></p>
<p>Certain data have a natural hierarchical structure. For example, say that the variables are City, Town, State, Country, and Continent. If we used hierarchical clustering, this pattern could be established even if we did not have labels for Cities, Towns, and so forth.</p>
<p>The <code>hclust()</code> function implements hierarchical clustering in R. In the following example; we use the data of previous section to plot the hierarchical clustering dendrogram using full, single, and average linkage clustering Euclidean distance as the dissimilarity measure. We begin by clustering observations using complete linkage. The <code>dist()</code> function is used
to compute the 50 <span class="math inline">\(\times\)</span> 50 inter-observation Euclidean distance matrix:</p>
<div class="sourceCode" id="cb460"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb460-1"><a href="unsupervised-learning.html#cb460-1" aria-hidden="true" tabindex="-1"></a>hc_complete <span class="ot">=</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(x), <span class="at">method =</span> <span class="st">&quot;complete&quot;</span>)</span></code></pre></div>
<p>We could just as easily perform hierarchical clustering with average or single linkage instead:</p>
<div class="sourceCode" id="cb461"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb461-1"><a href="unsupervised-learning.html#cb461-1" aria-hidden="true" tabindex="-1"></a>hc_average <span class="ot">=</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(x), <span class="at">method =</span> <span class="st">&quot;average&quot;</span>)</span>
<span id="cb461-2"><a href="unsupervised-learning.html#cb461-2" aria-hidden="true" tabindex="-1"></a>hc_single <span class="ot">=</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(x), <span class="at">method =</span> <span class="st">&quot;single&quot;</span>)</span></code></pre></div>
<p>We can now plot the dendrograms obtained using the usual <code>plot()</code> function.
The numbers at the bottom of the plot identify each observation:</p>
<div class="sourceCode" id="cb462"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb462-1"><a href="unsupervised-learning.html#cb462-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb462-2"><a href="unsupervised-learning.html#cb462-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc_complete,<span class="at">main =</span> <span class="st">&quot;Complete Linkage&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">sub =</span> <span class="st">&quot;&quot;</span>, <span class="at">cex =</span> .<span class="dv">9</span>)</span>
<span id="cb462-3"><a href="unsupervised-learning.html#cb462-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc_average, <span class="at">main =</span> <span class="st">&quot;Average Linkage&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">sub =</span> <span class="st">&quot;&quot;</span>, <span class="at">cex =</span> .<span class="dv">9</span>)</span>
<span id="cb462-4"><a href="unsupervised-learning.html#cb462-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc_single, <span class="at">main =</span> <span class="st">&quot;Single Linkage&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">sub =</span> <span class="st">&quot;&quot;</span>, <span class="at">cex =</span> .<span class="dv">9</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-278-1.png" width="576" /></p>
<p>To determine the cluster labels for each observation associated with a given cut of the dendrogram, we can use the <code>cutree()</code> function:</p>
<div class="sourceCode" id="cb463"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb463-1"><a href="unsupervised-learning.html#cb463-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cutree</span>(hc_complete, <span class="dv">2</span>)</span></code></pre></div>
<pre><code>##  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
## [45] 2 2 2 2 2 2</code></pre>
<div class="sourceCode" id="cb465"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb465-1"><a href="unsupervised-learning.html#cb465-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cutree</span>(hc_average, <span class="dv">2</span>)</span></code></pre></div>
<pre><code>##  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 1 2 2 2 2 2 2 2 2 2 2 1
## [45] 2 1 2 2 2 2</code></pre>
<div class="sourceCode" id="cb467"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb467-1"><a href="unsupervised-learning.html#cb467-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cutree</span>(hc_single, <span class="dv">2</span>)</span></code></pre></div>
<pre><code>##  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## [45] 1 1 1 1 1 1</code></pre>
<p>For this data, complete and average linkage generally separate the observations into their correct groups. However, single linkage identifies one observation as belonging to its own cluster. A more sensible answer is obtained when four clusters are selected, although there are still two singletons:</p>
<div class="sourceCode" id="cb469"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb469-1"><a href="unsupervised-learning.html#cb469-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cutree</span>(hc_single, <span class="dv">4</span>)</span></code></pre></div>
<pre><code>##  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 4 3 3
## [45] 3 3 3 3 3 3</code></pre>
<p>To scale the variables before performing hierarchical clustering of the
observations, we can use the <code>scale()</code> function:</p>
<div class="sourceCode" id="cb471"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb471-1"><a href="unsupervised-learning.html#cb471-1" aria-hidden="true" tabindex="-1"></a>xsc <span class="ot">=</span> <span class="fu">scale</span>(x)</span>
<span id="cb471-2"><a href="unsupervised-learning.html#cb471-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(<span class="fu">dist</span>(xsc), <span class="at">method =</span> <span class="st">&quot;complete&quot;</span>), <span class="at">main =</span> <span class="st">&quot;Hierarchical Clustering with Scaled Features&quot;</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-281-1.png" width="576" /></p>
<p>Correlation-based distance can be computed using the <code>as.dist()</code> function, which converts an arbitrary square symmetric matrix into a form that the <code>hclust()</code> function recognizes as a distance matrix. However, this only makes sense for data with <strong>at least three features</strong> since the absolute correlation between any two observations with measurements on two features is always 1. Let us generate and cluster a three-dimensional data set:</p>
<div class="sourceCode" id="cb472"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb472-1"><a href="unsupervised-learning.html#cb472-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(<span class="dv">30</span><span class="sc">*</span><span class="dv">3</span>), <span class="at">ncol =</span> <span class="dv">3</span>)</span>
<span id="cb472-2"><a href="unsupervised-learning.html#cb472-2" aria-hidden="true" tabindex="-1"></a>dd <span class="ot">=</span> <span class="fu">as.dist</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">cor</span>(<span class="fu">t</span>(x)))</span>
<span id="cb472-3"><a href="unsupervised-learning.html#cb472-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(dd, <span class="at">method =</span> <span class="st">&quot;complete&quot;</span>), <span class="at">main =</span> <span class="st">&quot;Complete Linkage with Correlation-Based Distance&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">sub =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-282-1.png" width="576" /></p>
<div id="example-clustering-cancel-cells" class="section level3" number="17.6.1">
<h3><span class="header-section-number">17.6.1</span> Example: Clustering Cancel Cells</h3>
<p>Unsupervised techniques are often used in the analysis of genomic data. In this example, we will see how hierarchical and K-means clustering compare on the <code>NCI60</code> cancer cell line microarray data, which consists of 6,830 gene expression measurements on 64 cancer cell lines:</p>
<div class="sourceCode" id="cb473"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb473-1"><a href="unsupervised-learning.html#cb473-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The NCI60 data</span></span>
<span id="cb473-2"><a href="unsupervised-learning.html#cb473-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ISLR)</span>
<span id="cb473-3"><a href="unsupervised-learning.html#cb473-3" aria-hidden="true" tabindex="-1"></a>nci_labels <span class="ot">=</span> NCI60<span class="sc">$</span>labs</span>
<span id="cb473-4"><a href="unsupervised-learning.html#cb473-4" aria-hidden="true" tabindex="-1"></a>nci_data <span class="ot">=</span> NCI60<span class="sc">$</span>data</span></code></pre></div>
<p>Each cell line is labeled with a cancer type. We will ignore the cancer types in performing clustering, as these are unsupervised techniques. After performing clustering, we will use this column to see the extent to which these cancer types agree with the results of these unsupervised techniques.</p>
<p>The data has 64 rows and 6,830 columns.</p>
<div class="sourceCode" id="cb474"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb474-1"><a href="unsupervised-learning.html#cb474-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(nci_data)</span></code></pre></div>
<pre><code>## [1]   64 6830</code></pre>
<p>Let‚Äôs take a look at the cancer types for the cell lines:</p>
<div class="sourceCode" id="cb476"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb476-1"><a href="unsupervised-learning.html#cb476-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(nci_labels)</span></code></pre></div>
<pre><code>## nci_labels
##      BREAST         CNS       COLON K562A-repro K562B-repro    LEUKEMIA MCF7A-repro 
##           7           5           7           1           1           6           1 
## MCF7D-repro    MELANOMA       NSCLC     OVARIAN    PROSTATE       RENAL     UNKNOWN 
##           1           8           9           6           2           9           1</code></pre>
<p>We now proceed to hierarchically cluster the cell lines in the <code>NCI60</code> data to determine whether or not the observations cluster into distinct types of cancer. To begin, we standardize the variables to have mean zero and standard deviation one. This step is optional and needs only be performed if we want each gene to be on the same scale:</p>
<div class="sourceCode" id="cb478"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb478-1"><a href="unsupervised-learning.html#cb478-1" aria-hidden="true" tabindex="-1"></a>sd_data <span class="ot">=</span> <span class="fu">scale</span>(nci_data)</span></code></pre></div>
<p>We now perform hierarchical clustering of the observations using complete, single, and average linkage. We will use standard Euclidean distance as the dissimilarity measure:</p>
<div class="sourceCode" id="cb479"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb479-1"><a href="unsupervised-learning.html#cb479-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb479-2"><a href="unsupervised-learning.html#cb479-2" aria-hidden="true" tabindex="-1"></a>data_dist <span class="ot">=</span> <span class="fu">dist</span>(sd_data)</span>
<span id="cb479-3"><a href="unsupervised-learning.html#cb479-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(data_dist), <span class="at">labels =</span> nci_labels, <span class="at">main =</span> <span class="st">&quot;Complete Linkage&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">sub =</span> <span class="st">&quot;&quot;</span>,<span class="at">ylab =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb479-4"><a href="unsupervised-learning.html#cb479-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(data_dist, <span class="at">method =</span> <span class="st">&quot;average&quot;</span>), <span class="at">labels =</span> nci_labels, <span class="at">main =</span> <span class="st">&quot;Average Linkage&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">sub =</span> <span class="st">&quot;&quot;</span>,<span class="at">ylab =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb479-5"><a href="unsupervised-learning.html#cb479-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(data_dist, <span class="at">method =</span> <span class="st">&quot;single&quot;</span>), <span class="at">labels =</span> nci_labels,  <span class="at">main =</span> <span class="st">&quot;Single Linkage&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">sub =</span> <span class="st">&quot;&quot;</span>,<span class="at">ylab =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-287-1.png" width="576" /></p>
<p>We see that the choice of linkage certainly does affect the results obtained. Typically, a single linkage will tend to yield trailing clusters: very large clusters onto which individual observations attach one-by-one. On the other hand, complete and average linkage tends to yield more balanced, attractive clusters. For this reason, the complete and average linkage is generally preferred to single linkage. Cell lines within a single cancer type tend to cluster together, although the clustering is not perfect.</p>
<p>Let us use our complete linkage hierarchical clustering for the analysis. We can cut the dendrogram at the height that will yield a particular number of clusters, say 4:</p>
<div class="sourceCode" id="cb480"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb480-1"><a href="unsupervised-learning.html#cb480-1" aria-hidden="true" tabindex="-1"></a>hc_out <span class="ot">=</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(sd_data))</span>
<span id="cb480-2"><a href="unsupervised-learning.html#cb480-2" aria-hidden="true" tabindex="-1"></a>hc_clusters <span class="ot">=</span> <span class="fu">cutree</span>(hc_out,<span class="dv">4</span>)</span>
<span id="cb480-3"><a href="unsupervised-learning.html#cb480-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(hc_clusters,nci_labels)</span></code></pre></div>
<pre><code>##            nci_labels
## hc_clusters BREAST CNS COLON K562A-repro K562B-repro LEUKEMIA MCF7A-repro MCF7D-repro
##           1      2   3     2           0           0        0           0           0
##           2      3   2     0           0           0        0           0           0
##           3      0   0     0           1           1        6           0           0
##           4      2   0     5           0           0        0           1           1
##            nci_labels
## hc_clusters MELANOMA NSCLC OVARIAN PROSTATE RENAL UNKNOWN
##           1        8     8       6        2     8       1
##           2        0     1       0        0     1       0
##           3        0     0       0        0     0       0
##           4        0     0       0        0     0       0</code></pre>
<p>There are some clear patterns. All the leukemia cell lines fall in cluster 3, while the breast cancer cell lines are spread out over three different clusters. We can plot the cut on the dendrogram that produces these four clusters using the <code>abline()</code> function, which draws a straight line on top of any existing plot in R:</p>
<div class="sourceCode" id="cb482"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb482-1"><a href="unsupervised-learning.html#cb482-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb482-2"><a href="unsupervised-learning.html#cb482-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc_out, <span class="at">labels =</span> nci_labels)</span>
<span id="cb482-3"><a href="unsupervised-learning.html#cb482-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">139</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="Exam-PA-Study-Manual_files/figure-html/unnamed-chunk-289-1.png" width="576" /></p>
<p>Printing the output of <code>hclust</code> gives a useful brief summary of the object:</p>
<div class="sourceCode" id="cb483"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb483-1"><a href="unsupervised-learning.html#cb483-1" aria-hidden="true" tabindex="-1"></a>hc_out</span></code></pre></div>
<pre><code>## 
## Call:
## hclust(d = dist(sd_data))
## 
## Cluster method   : complete 
## Distance         : euclidean 
## Number of objects: 64</code></pre>
<p>We claimed earlier that K-means clustering and hierarchical clustering with the dendrogram cut to obtain the same number of clusters could yield <strong>very</strong> different results. How do these <code>NCI60</code> hierarchical clustering results compare to what we get if we perform K-means clustering
with <code>K  =  4</code>?</p>
<div class="sourceCode" id="cb485"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb485-1"><a href="unsupervised-learning.html#cb485-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb485-2"><a href="unsupervised-learning.html#cb485-2" aria-hidden="true" tabindex="-1"></a>km_out <span class="ot">=</span> <span class="fu">kmeans</span>(sd_data, <span class="dv">4</span>, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb485-3"><a href="unsupervised-learning.html#cb485-3" aria-hidden="true" tabindex="-1"></a>km_clusters <span class="ot">=</span> km_out<span class="sc">$</span>cluster</span></code></pre></div>
<p>We can use a confusion matrix to compare the differences in how the two methods assigned observations to clusters:</p>
<div class="sourceCode" id="cb486"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb486-1"><a href="unsupervised-learning.html#cb486-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(km_clusters,hc_clusters)</span></code></pre></div>
<pre><code>##            hc_clusters
## km_clusters  1  2  3  4
##           1 11  0  0  9
##           2 20  7  0  0
##           3  9  0  0  0
##           4  0  0  8  0</code></pre>
<p>We see that the four clusters obtained using hierarchical clustering and Kmeans clustering are somewhat different. Cluster 2 in K-means clustering is identical to cluster 3 in hierarchical clustering. However, the other clusters differ: for instance, cluster 4 in K-means clustering contains a portion of the observations assigned to cluster 1 by hierarchical clustering and all of the observations assigned to cluster 2 by hierarchical clustering.</p>
</div>
<div id="references" class="section level3" number="17.6.2">
<h3><span class="header-section-number">17.6.2</span> References</h3>
<p>These examples are an adaptation of p.¬†404-407, 410-413 of ‚ÄúIntroduction to Statistical Learning with Applications in R‚Äù by Gareth James, Daniela Witten, Trevor Hastie, and Robert Tibshirani. Adapted by R. Jordan Crouser at Smith College for SDS293: Machine Learning (Spring 2016), and re-implemented in Fall 2016 <code>tidyverse</code> format by Amelia McNamara and R. Jordan Crouser at Smith College.</p>
<p>Used with permission from Jordan Crouser at Smith College and to the following contributors on GitHub:</p>
<ul>
<li>github.com/jcrouser</li>
<li>github.com/AmeliaMN</li>
<li>github.com/mhusseinmidd</li>
<li>github.com/rudeboybert</li>
<li>github.com/ijlyttle</li>
</ul>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="tree-based-models.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="writing-and-communication.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/sdcastillo/PA-R-Study-Manual/edit/master/06-unsupervised-learning.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Exam-PA-Study-Manual.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
